### PR TITLE
OMX and IMX HWDecRenderer

### PR DESCRIPTION
**Pull request by Memphiz**
This refactors OMX and IMX HWRendering out into its own classes. 

This is not compile time tested.

@fritsch (IMX) and @popcornmix (OMX) for any needed compile fixes and runtime tests if possible.

That stagefright and android stuff might be a bit over my top with only a text editor. Also my time is sucked down again now. Maybe @koying can find some time at some point to do that android work.

Also i tried to not break the other renderers (eglimg and stuff). But i also can't compile / runtime test that.

### CODE REVIEW

**Comment by fritsch** [comment]
@smallint @wolfgar please have a look - if we want to invest this work into the IMX platform.

**Comment by Memphiz** [comment]
Its only moving code around ... should not have any effect on runtime [tm] ;D

**Comment by fritsch** [comment]
@Memphiz thanks very much for starting this

**Comment by popcornmix** [comment]
It's okay on Pi, with a few typos fixed. See: https://github.com/popcornmix/xbmc/commit/6682ecece0aca418d6371c2d4e1109771f1e2630

**Comment by Memphiz** [comment]
Nice - thx - squashed your changes in...

**Comment by Memphiz** [comment]
I think i do stagefright and mediacodec as well ...

**Comment by fritsch** [comment]
@Memphiz thanks very much in advance. But be careful those are more hard to do as they use global callbacks.

**Comment by Memphiz** [comment]
@fritsch - do you mean this? 

https://github.com/FernetMenta/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.cpp#L74

strangly this is nowhere used really ... *wonders*

**Comment by fritsch** [comment]
No afaik those callbacks are all in BaseRender.cpp where they should not be at all ...

https://github.com/FernetMenta/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp#L58

**Comment by Memphiz** [comment]
Humm - well i think this still works ... added mediacodec and stagefright - those won't work maybe - figuring out the missing bits is up to others imo - will dive back into my real life for a while now ;)

**Comment by wolfgar** [comment]
It looks like a nice move for sure !
I will review in details, compile, test quickly and report back
regards

**Comment by koying** [comment]
@Memphiz Thanks a lot for the helping hand :)

**Comment by Memphiz** [comment]
@koying added the missing mediacodec impl

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/Makefile.in [hunk 1]
```diff
@@ -3,11 +3,21 @@ SRCS=
 ifeq (@USE_OPENGL@,1)
 SRCS += RendererVAAPI.cpp
 SRCS += RendererVDPAU.cpp
+SRCS += RendererStagefright.cpp
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/Makefile.in]: those can't be used with OpenGL, right?

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVTB.cpp [hunk 2]
```diff
@@ -160,6 +183,11 @@ bool CRendererVTB::RenderHook(int index)
   return true;
 }
 
+void CRendererVTB::AddSupportedHwRenderFormats()
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVTB.cpp]: those are not needed anymore. the old code had a static renderer and decoder needed to adapt to renderer. now it is the other way round. the correct type of renderer is chosen and instantiated on ERenderFormat set be decoder.
https://github.com/FernetMenta/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp#L653

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/Makefile.in (hunk 1 above)
Comment (Memphiz) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/Makefile.in]: I was not sure if android-x86 was opengl or gles. If the latter i they need to go here - yep.

File: xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVTB.cpp (hunk 2 above)
Comment (Memphiz) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVTB.cpp]: right - will remove

Comment (Memphiz) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVTB.cpp]: Mhhh are you sure? m_formats is used in "GetRenderInfo" for example. If the renderer doesn't add that format to it - who does instead? What you stated would mean i could remove m_formats completely - but i don't see how.

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVTB.cpp]: Yes, I am sure. At the time a decoder is crated, there is no instance of a renderer and m_format may be empty anyway. I will drop m_formats later.

**Comment by koying** [comment]
This rewrite is a major incentive to finally drop stagefright for android, so nobody should spend time on that. 

@Memphiz sorry for loosing your time. As far as I'm concerned, you can drop it altogether.

**Comment by Memphiz** [comment]
dropped stagefright and AddSupportedHwRenderFormats

**Comment by Memphiz** [comment]
was this a pre mature merge or did you get a working indication over different channels? Orare you gonna sqaushdown properly anyways at some point?

**Comment by FernetMenta** [comment]
just a premature merge before rebase on mainline. does not harm to integrate more often here, right?

**Comment by Memphiz** [comment]
You know what? I somehow forgot about the fact that this isn't mainline :D - sure integrate as often as you feel comfortable ;)

**Comment by wolfgar** [comment]
Hi there 
I have begun to fix  compilation for imx6 here https://github.com/wolfgar/xbmc/commits/hwdecrenderer
For now I have a strange error regarding locking from glibc that triggers a few seconds after boot so I cannot test properly 
I will investigate my issue and continue testing properly asap

**Comment by FernetMenta** [comment]
thanks for update. highly appreciated.

**Comment by wolfgar** [comment]
Hi,
Sorry for long delay, I have now tested and fixed issues I have been faced with...
The https://github.com/wolfgar/xbmc/commits/hwdecrenderer branch is now able to play fine on imx6
Regards

**Comment by FernetMenta** [comment]
@wolfgar could you please PR the changes to my repo?

**Comment by wolfgar** [comment]
of course yes, done in PR #300

**Comment by fritsch** [comment]
@codesnake: We still search a volunteer that ports AMLogic to this new way of Codec / Player architecture. @stefansaraev told me that you currently work on AML changes - could you base your work on these changes here that will be future kodi?

**Comment by codesnake** [comment]
@fritsch What you mean by new way of Codec / Player architecture? In case of Amlogic we don't have any control over rendring of video picture. Rendering is performed by aml hw/drivers.

**Comment by codesnake** [comment]
@fritsch I want to help just don't understand the meaning of these new changes.

**Comment by fritsch** [comment]
@codesnake: Thank you very much, very welcome. As AMLogic is one of the BYPASS renderers I asked @fernetMenta to summarize the steps that need to be done as I might get some bits wrong and don't want to guide you into a wrong direction.

**Comment by codesnake** [comment]
@fritsch Thanks!

**Comment by FernetMenta** [comment]
@codesnake have you seen this section on the forum? http://forum.kodi.tv/forumdisplay.php?fid=240

**Comment by codesnake** [comment]
@FernetMenta No I have not seen, but I'll read it and subscribe. Thanks!

**Pull request merged**